File: ZipFlash_debugBuild/app/src/main/java/com/zipflash/mrp/helper/ScriptRunner.java
package com.zipflash.mrp.helper;

import android.content.Context;
import android.net.Uri;
import android.widget.TextView;

import com.zipflash.mrp.SettingsHelper;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.File;

import rikka.shizuku.Shizuku;
import rikka.shizuku.ShizukuRemoteProcess;

public class ScriptRunner {

    public interface OnScriptFinishedListener {
        void onFinished();
        void onError(String error);
    }

    public static void runSingleSh(final Context context,
                                   final Uri uri,
                                   final TextView outputView,
                                   final OnScriptFinishedListener listener) {

        new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        final String fileName = FileHelper.getFileName(context, uri);
                        if (fileName == null) {
                            outputView.post(new Runnable() {
                                    @Override
                                    public void run() {
                                        outputView.append("[!] Invalid file.\n");
                                    }
                                });
                            if (listener != null) listener.onError("Invalid file.");
                            return;
                        }

                        final String targetDir = "/data/local/tmp/MRP";
                        final String targetPath = targetDir + "/" + fileName;

                        // Copy file into /data/local/tmp/MRP
                        final ShizukuRemoteProcess copyProcess = Shizuku.newProcess(
                            new String[]{"sh", "-c", "mkdir -p " + targetDir + " && cat > " + targetPath},
                            null,
                            null
                        );

                        byte[] buffer = new byte[4096];
                        InputStream is = context.getContentResolver().openInputStream(uri);
                        OutputStream os = copyProcess.getOutputStream();
                        int len;
                        while ((len = is.read(buffer)) > 0) {
                            os.write(buffer, 0, len);
                        }
                        os.close();
                        is.close();
                        copyProcess.waitFor();
                        copyProcess.destroy();

                        // Make file executable
                        Shizuku.newProcess(
                            new String[]{"chmod", "+x", targetPath},
                            null,
                            null
                        ).waitFor();

                        // ✅ Hand off execution to ShellHelper
                        ShellHelper.runShellScript(context, targetPath, outputView, new ShellHelper.OnScriptFinishedListener() {
                                @Override
                                public void onFinished() {
                                    if (listener != null) listener.onFinished();
                                }

                                @Override
                                public void onError(String error) {
                                    if (listener != null) listener.onError(error);
                                }
                            });

                    } catch (final Exception e) {
                        outputView.post(new Runnable() {
                                @Override
                                public void run() {
                                    outputView.append("[!] Failed to prepare file: " + e.getMessage() + "\n");
                                }
                            });
                        if (listener != null) listener.onError(e.getMessage());
                    }
                }
            }).start();
    }
}

File: ZipFlash/app/src/main/java/com/zipflash/mrp/helper/ScriptRunner.kt
package com.zipflash.mrp.helper

import android.content.Context
import android.net.Uri
import android.widget.TextView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import rikka.shizuku.Shizuku
import java.io.File

object ScriptRunner {
    interface OnScriptFinishedListener {
        fun onFinished()
        fun onError(error: String)
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun runSingleShBlocking(context: Context, uri: Uri, listener: OnScriptFinishedListener?) {
        try {
            if (!CheckPermHelper.hasShizukuPermission()) {
                listener?.onError("Shizuku permission not available.")
                return
            }

            val fileName = FileHelper.getFileName(context, uri)
            if (fileName == null) {
                listener?.onError("Invalid file.")
                return
            }

            val targetDir = "/data/local/tmp/MRP"
            val targetPath = "$targetDir/$fileName"

            val copyProcess = Shizuku.newProcess(
                arrayOf("sh", "-c", "mkdir -p $targetDir && cat > $targetPath"), null, null
            )

            val `is` = context.contentResolver.openInputStream(uri) ?: throw Exception("Failed to open input stream")
            val os = copyProcess.outputStream
            val buffer = ByteArray(4096)
            var len: Int
            while (`is`.read(buffer).also { len = it } > 0) {
                os.write(buffer, 0, len)
            }
            os.close()
            `is`.close()
            copyProcess.waitFor()
            copyProcess.destroy()

            Shizuku.newProcess(arrayOf("chmod", "+x", targetPath), null, null).waitFor()

            ShellHelper.runShellScriptBlocking(context, targetPath, object : ShellHelper.OnScriptFinishedListener {
                override fun onFinished() {
                    listener?.onFinished()
                }

                override fun onError(error: String) {
                    listener?.onError(error)
                }
            })
        } catch (e: Exception) {
            listener?.onError(e.message ?: "Unknown error")
        }
    }

    fun runSingleSh(context: Context, uri: Uri, outputView: TextView, listener: OnScriptFinishedListener?) {
        scope.launch {
            try {
                if (!CheckPermHelper.hasShizukuPermission()) {
                    outputView.post { outputView.append("[!] Shizuku permission not available.\n") }
                    listener?.onError("Shizuku permission not available.")
                    return@launch
                }

                val fileName = FileHelper.getFileName(context, uri)
                if (fileName == null) {
                    outputView.post { outputView.append("[!] Invalid file.\n") }
                    listener?.onError("Invalid file.")
                    return@launch
                }

                val targetDir = "/data/local/tmp/MRP"
                val targetPath = "$targetDir/$fileName"

                val copyProcess = Shizuku.newProcess(
                    arrayOf("sh", "-c", "mkdir -p $targetDir && cat > $targetPath"), null, null
                )

                val `is` = context.contentResolver.openInputStream(uri) ?: throw Exception("Failed to open input stream")
                val os = copyProcess.outputStream
                val buffer = ByteArray(4096)
                var len: Int
                while (`is`.read(buffer).also { len = it } > 0) {
                    os.write(buffer, 0, len)
                }
                os.close()
                `is`.close()
                copyProcess.waitFor()
                copyProcess.destroy()

                Shizuku.newProcess(arrayOf("chmod", "+x", targetPath), null, null).waitFor()

                ShellHelper.runShellScript(context, targetPath, outputView, object : ShellHelper.OnScriptFinishedListener {
                    override fun onFinished() {
                        listener?.onFinished()
                    }

                    override fun onError(error: String) {
                        listener?.onError(error)
                    }
                })

            } catch (e: Exception) {
                outputView.post { outputView.append("[!] Failed to prepare file: ${e.message}\n") }
                listener?.onError(e.message ?: "Unknown error")
            }
        }
    }
}
